File: Network Packet Routing Simulator/dijkstra.cpp
/*
Dijkstra's Algorithm
Single Pair Shortest Path
T.C: O((V + E)logV)
S.C: O(V + E)
*/

#include <bits/stdc++.h>
#include <windows.h>

using namespace std;

#define MX 105
#define INF 1e8

struct node
{
    int node;
    int cost;
};

vector<node> Graph[MX];
bool vis[MX];
int dist[MX];
int parent[MX];

void reset()
{
    for (int i = 0; i < MX; i++)
    {
        Graph[i].clear();
        vis[i] = false;
        dist[i] = INF;
        parent[i] = -1;
    }
}

class cmp
{
public:
    bool operator()(node &A, node &B)
    {
        return A.cost > B.cost;
    }
};

void dijkstra(int source)
{
    priority_queue<node, vector<node>, cmp> PQ;
    dist[source] = 0;
    PQ.push({source, 0});

    cout << "\nStarting Dijkstra from node " << source << "...\n";

    while (!PQ.empty())
    {
        node current = PQ.top();
        PQ.pop();

        int u = current.node;
        int cost = current.cost;

        if (vis[u])
            continue;

        vis[u] = true;
        cout << "Visiting node " << u << " with current cost " << cost << "\n";

        for (auto &edge : Graph[u])
        {
            int v = edge.node;
            int w = edge.cost;

            if (!vis[v] && dist[u] + w < dist[v])
            {
                dist[v] = dist[u] + w;
                parent[v] = u;
                PQ.push({v, dist[v]});
                cout << "  Updated distance of node " << v << " to " << dist[v] << " via node " << u << "\n";
            }
        }
    }
}

vector<int> reconstructPath(int src, int dest)
{
    vector<int> path;
    for (int at = dest; at != -1; at = parent[at])
        path.push_back(at);
    reverse(path.begin(), path.end());
    if (path[0] == src)
        return path;
    return {};
}

void simulatePacketMovement(const vector<int> &path)
{
    cout << "\nSimulating packet movement from source to destination...\n";
    for (int i = 0; i < path.size(); i++)
    {
        cout << "Packet at router " << path[i];
        if (i != path.size() - 1)
            cout << " -> moving to router " << path[i + 1];
        cout << "\n";

        Sleep(700);
    }
    cout << "Packet reached destination.\n";
}

int main()
{
    reset();

    int nodes, edges;
    cout << "Enter number of routers (nodes) and links (edges): ";
    cin >> nodes >> edges;

    cout << "Enter edges in format: from to weight\n";
    for (int i = 1; i <= edges; i++)
    {
        int u, v, w;
        cin >> u >> v >> w;
        Graph[u].push_back({v, w});
        Graph[v].push_back({u, w});
    }

    int source, destination;
    cout << "Enter source and destination routers: ";
    cin >> source >> destination;

    dijkstra(source);

    cout << "\n--- Final Distances from Source " << source << " ---\n";
    for (int i = 1; i <= nodes; i++)
    {
        cout << "Router " << i << ": ";
        if (dist[i] == INF)
            cout << "UNREACHABLE\n";
        else
            cout << dist[i] << "\n";
    }

    vector<int> path = reconstructPath(source, destination);
    if (path.empty())
    {
        cout << "\nNo path found from " << source << " to " << destination << ".\n";
    }
    else
    {
        cout << "\nShortest Path: ";
        for (int i = 0; i < path.size(); i++)
        {
            cout << path[i];
            if (i != path.size() - 1)
                cout << " -> ";
        }
        cout << "\n";

        simulatePacketMovement(path);
    }

    return 0;
}

/*
Sample Input:
5 6
1 2 2
1 3 1
2 3 1
1 4 3
4 5 2
5 3 5
1 5
*/

File: Network Packet Routing Simulator/dijkstra.py
import tkinter as tk
from tkinter import messagebox
import heapq
import time
import threading
import math

class NetworkSimulator:
    def __init__(self, master):
        self.master = master
        master.title("Network Packet Routing Simulator (Dijkstra)")

        # Instructions label
        self.instructions = tk.Label(master, text="Enter input (example format):\n"
                                                  "5 6\n"
                                                  "1 2 2\n"
                                                  "1 3 1\n"
                                                  "2 3 1\n"
                                                  "1 4 3\n"
                                                  "4 5 2\n"
                                                  "5 3 5\n"
                                                  "1 5", font=("Arial", 10))
        self.instructions.pack()

        # Multiline input box
        self.input_text = tk.Text(master, height=10, width=40)
        self.input_text.pack()

        # Run button
        self.run_button = tk.Button(master, text="Build Graph & Find Path", command=self.run_simulation)
        self.run_button.pack()

        # Canvas for drawing
        self.canvas = tk.Canvas(master, width=700, height=500, bg="white")
        self.canvas.pack()

        # Result label
        self.result_label = tk.Label(master, text="", font=("Arial", 14))
        self.result_label.pack()

        # To store node positions and canvas objects
        self.node_positions = {}
        self.node_objects = {}
        self.edge_objects = []

    def parse_input(self, input_str):
        lines = input_str.strip().split('\n')
        if len(lines) < 2:
            messagebox.showerror("Error", "Invalid input format")
            return None

        try:
            nodes, edges = map(int, lines[0].split())
            edge_lines = lines[1:1 + edges]
            last_line = lines[1 + edges]
        except Exception:
            messagebox.showerror("Error", "Invalid input format")
            return None

        graph = {i: {} for i in range(1, nodes + 1)}

        for line in edge_lines:
            u, v, w = map(int, line.split())
            graph[u][v] = w
            graph[v][u] = w  # Assuming undirected graph

        source, destination = map(int, last_line.split())

        return nodes, graph, source, destination

    def generate_positions(self, nodes):
        # Arrange nodes evenly in a circle
        center_x, center_y = 350, 250
        radius = 200
        self.node_positions = {}
        for i in range(1, nodes + 1):
            angle = 2 * math.pi * (i - 1) / nodes
            x = center_x + int(radius * math.cos(angle))
            y = center_y + int(radius * math.sin(angle))
            self.node_positions[i] = (x, y)

    def draw_graph(self, graph):
        self.canvas.delete("all")
        self.node_objects.clear()
        self.edge_objects.clear()

        # Draw edges with weights
        drawn_edges = set()
        for u in graph:
            for v in graph[u]:
                if (v, u) in drawn_edges:
                    continue
                x1, y1 = self.node_positions[u]
                x2, y2 = self.node_positions[v]
                self.canvas.create_line(x1, y1, x2, y2, fill="gray", width=2)
                mid_x, mid_y = (x1 + x2)//2, (y1 + y2)//2
                self.canvas.create_text(mid_x, mid_y, text=str(graph[u][v]), font=("Arial", 10, "bold"), fill="blue")
                drawn_edges.add((u, v))

        # Draw nodes
        for node, (x, y) in self.node_positions.items():
            oval = self.canvas.create_oval(x - 20, y - 20, x + 20, y + 20, fill="lightblue", outline="black", width=2)
            self.canvas.create_text(x, y, text=str(node), font=("Arial", 12, "bold"))
            self.node_objects[node] = oval

    def dijkstra(self, graph, start, goal):
        queue = [(0, start)]
        dist = {node: float('inf') for node in graph}
        dist[start] = 0
        parent = {node: None for node in graph}
        visited = set()

        while queue:
            cost, node = heapq.heappop(queue)
            if node in visited:
                continue
            visited.add(node)
            if node == goal:
                break
            for neighbor in graph[node]:
                if neighbor not in visited:
                    new_cost = cost + graph[node][neighbor]
                    if new_cost < dist[neighbor]:
                        dist[neighbor] = new_cost
                        parent[neighbor] = node
                        heapq.heappush(queue, (new_cost, neighbor))

        # Reconstruct path
        path = []
        cur = goal
        while cur is not None:
            path.append(cur)
            cur = parent[cur]
        path.reverse()

        if path[0] == start:
            return path, dist[goal]
        else:
            return [], float('inf')

    def animate_path(self, path):
        # Highlight nodes and edges on path step-by-step
        for i in range(len(path)):
            node = path[i]
            # Highlight node
            self.canvas.itemconfig(self.node_objects[node], fill="orange")
            if i > 0:
                # Highlight edge from previous node
                prev = path[i-1]
                x1, y1 = self.node_positions[prev]
                x2, y2 = self.node_positions[node]
                self.canvas.create_line(x1, y1, x2, y2, fill="red", width=4, tags="highlight")
            self.canvas.update()
            time.sleep(0.7)

    def run_simulation(self):
        input_str = self.input_text.get("1.0", tk.END)
        parsed = self.parse_input(input_str)
        if not parsed:
            return
        nodes, graph, source, destination = parsed

        self.generate_positions(nodes)
        self.draw_graph(graph)
        self.result_label.config(text="Calculating shortest path...")

        # Run Dijkstra and animation in a separate thread to prevent UI freeze
        def run():
            path, cost = self.dijkstra(graph, source, destination)
            if not path:
                self.result_label.config(text=f"No path from {source} to {destination}")
                return
            self.result_label.config(text=f"Shortest Path: {' → '.join(map(str,path))} | Total Cost: {cost}")
            self.animate_path(path)

        threading.Thread(target=run).start()


if __name__ == "__main__":
    root = tk.Tk()
    app = NetworkSimulator(root)
    root.mainloop()
